vllm/model_executor/models/molmo.py

```diff
@@ -0,0 +1,1130 @@
+        """
+        B, T, N, D = images.shape
+
+        mask = ~torch.all(images.view(B * T, N, D) == -1, dim=(1, 2), keepdim=True)
+
+        images = images.view(B * T, N, D)
+        image_features = self.image_vit(images)
+
+        if self.vit_layers is not None:
+            features = []
+            for layer in self.vit_layers:
+                features.append(image_features[layer])
+            image_features = torch.cat(features, dim=-1)
+        else:
+            image_features = image_features[-1]
+
+        cls_embed: torch.Tensor = None
+        if self.num_prefix_tokens > 0:
+            cls_embed = image_features[:, 0]
+            image_features = image_features[:, 1:]
+
+        image_features = image_features * mask
+        image_features = image_features.view(B, T, N, -1)
+
+        cls_embed = cls_embed.view(B, T, -1) if cls_embed is not None else None
+
+        return image_features, cls_embed
+
+    def forward(self, images: torch.Tensor, image_masks: torch.Tensor) -> Tuple[torch.Tensor, Optional[torch.Tensor]]:
+
+        # image_features: (batch_size, num_crops(=num_image), num_patch, nximage_emb_dim)
+        batch_size, num_image = images.shape[:2]
+        image_features, cls_embed = self.encode_image(images)
+        
+        og_dtype = image_features.dtype
+        assert image_masks is not None
+        pad_embed = self.pad_embed[:, None, None, None, :]
+        all_pad = image_masks == 0
+        partial_pad = torch.logical_and(image_masks < 1, torch.logical_not(all_pad)).to(dtype=torch.float32)
+        all_pad = all_pad.to(dtype=torch.float32)
+        image_features = image_features + pad_embed[0] * torch.unsqueeze(all_pad, -1)
+        image_features = image_features + pad_embed[1] * torch.unsqueeze(partial_pad, -1)
+
+        image_features = image_features.to(og_dtype)
+
+        image_features = image_features.reshape(
+            (batch_size, num_image) + self.image_num_patch + (-1,),
+        )
+
+        if self.image_num_patch[0] % 2 == 1:
+            # Pad so we can still pool 2x2 patches
+            image_features = F.pad(
+                image_features,
+                (0, 0, 0, 1, 0, 1, 0, 0, 0, 0),
+            )
+
+        # image pooling
+        image_features = rearrange(
+            image_features,
+            'b n (h dh) (w dw) c -> (b n h w) (dh dw) c',
+            dh=2,
+            dw=2,
+        )
+
+        query = image_features.mean(-2, keepdim=True)
+        image_features = self.image_pooling_2d(query, image_features)
+
+        h, w = self.llm_patches_per_crop
+        image_features = image_features.view(batch_size, num_image, h * w, -1)
+
+        image_features = self.image_projector(image_features)
+
+        # image_features: (batch_size, num_image, num_patch, d_model)
+        # cls_embed: (batch_size, num_image, d_model)
+        return image_features, cls_embed
+
+
+class MolmoModel(nn.Module):
```

Comment:
I wonder if it's possible to reuse the existing `olmo`/`qwen2` implementation in vLLM. Because these backbones used in `molmo` have been implemented in vLLM.

I think we can clean up the code and prevent duplicate implementation if this is possible.

Reply:
We actually did that for Molmo models with Qwen2 backbone at first and it worked. However, for the model with Olmo backbone, we weren’t entirely sure if the vllm's Olmo model is similar to the newest Olmo version we're using--there were some slight architectural changes if I'm not wrong.  Therefore we decided to copy and rewrite everything in this single file for both backbones.

Reply:
Thanks for the explanation! The current `olmo` implementation in vLLM is based on HF format, so it might not be compatible with `molmo` which uses original format with slight difference architecture change.

Since Qwen2 backbone is working, I would prefer only porting the new architecture `olmo` backbone in this file rather than copying the `MolmoModel` which coupled `qwen2` and `olmo` backbone implementation. So that we can also migrate/update the current olmo implementation once the new olmo model is released.

With this, we can just simplify the implementation with arbitrary backbone like:
```python3
class MolmoForCausalLM(...):

    def __init__(...):
        ...

        if llm_backbone == "qwen2":
            self.language_model = Qwen2Model(...)
        elif llm_backbone == "olmo"
            self.language_model = NewOlmoModel(...)
```